packages/automerge-repo/src/storage/StorageSubsystem.ts

```diff
@@ -1,46 +1,148 @@
 import * as A from "@automerge/automerge/next"
 import debug from "debug"
-import * as sha256 from "fast-sha256"
 import { headsAreSame } from "../helpers/headsAreSame.js"
 import { mergeArrays } from "../helpers/mergeArrays.js"
 import { type DocumentId } from "../types.js"
-import { StorageAdapter, StorageKey } from "./StorageAdapter.js"
-
-// Metadata about a chunk of data loaded from storage. This is stored on the
-// StorageSubsystem so when we are compacting we know what chunks we can safely delete
-type StorageChunkInfo = {
-  key: StorageKey
-  type: ChunkType
-  size: number
-}
-
-export type ChunkType = "snapshot" | "incremental"
+import { StorageAdapter } from "./StorageAdapter.js"
+import { ChunkInfo, StorageKey } from "./types.js"
+import { keyHash, headsHash } from "./keyHash.js"
+import { chunkTypeFromKey } from "./chunkTypeFromKey.js"
+
+/**
+ * The storage subsystem is responsible for saving and loading Automerge documents to and from
+ * storage adapter. It also provides a generic key/value storage interface for other uses.
+ */
+export class StorageSubsystem {
+  /** Record of the latest heads we've loaded or saved for each document  */
+  #storedHeads: Map<DocumentId, A.Heads> = new Map()
 
-function keyHash(binary: Uint8Array) {
-  const hash = sha256.hash(binary)
-  const hashArray = Array.from(new Uint8Array(hash)) // convert buffer to byte array
-  const hashHex = hashArray.map(b => ("00" + b.toString(16)).slice(-2)).join("") // convert bytes to hex string
-  return hashHex
-}
+  /** Metadata on the chunks we've already loaded for each document */
+  #chunkInfos: Map<DocumentId, ChunkInfo[]> = new Map()
 
-function headsHash(heads: A.Heads): string {
-  const encoder = new TextEncoder()
-  const headsbinary = mergeArrays(heads.map((h: string) => encoder.encode(h)))
-  return keyHash(headsbinary)
-}
+  /** Flag to avoid compacting when a compaction is already underway */
+  #compacting = false
```

Comment:
can't prevent another tab from doing this... still probably a good idea but worth testing concurrent compactions

Reply:
yeah - maybe create a separate issue to track that?